lib/app/utils/hl7_validator.rb

```diff
@@ -19,23 +24,32 @@ def validate(resource, fhir_models_klass, profile_url = nil)
 
       result = RestClient.post "#{@validator_url}/validate", resource.source_contents, params: { profile: profile_url }
       outcome = fhir_models_klass.from_contents(result.body)
-      fatals = issues_by_severity(outcome.issue, 'fatal')
-      errors = issues_by_severity(outcome.issue, 'error')
-      warnings = issues_by_severity(outcome.issue, 'warning')
-      information = issues_by_severity(outcome.issue, 'information')
-      {
-        errors: fatals.concat(errors).reject(&:empty?),
-        warnings: warnings,
-        information: information
-      }
+
+      issues_by_severity(outcome.issue)
     end
 
     private
 
-    def issues_by_severity(issues, severity)
-      issues.reject { |i| i.code == 'code-invalid' }
-        .select { |i| i.severity == severity }
-        .map { |iss| "#{issue_location(iss)}: #{iss&.details&.text}" }
+    def issues_by_severity(issues)
+      errors = []
+      warnings = []
+      information = []
+
+      issues.each do |iss|
+        if iss.severity == 'information' || iss.code == 'code-invalid' || ISSUE_DETAILS_FILTER.any? { |filter| filter.match?(iss&.details&.text) }
+          information << "#{issue_location(iss)}: #{iss&.details&.text}"
+        elsif iss.severity == 'warning'
+          warnings << "#{issue_location(iss)}: #{iss&.details&.text}"
+        else
+          errors << "#{issue_location(iss)}: #{iss&.details&.text}"
+        end
+      end
+
+      {
+        errors: errors,
+        warnings: warnings,
+        information: information
+      }
```

Comment:
We can cut this method down in length and make it more readable (IMO) by using `#each_with_object`

```suggestion
      issues_hash = { errors: [], warnings: [], information: [] }
      issue_map = { fatal: :errors, error: :errors, warning: :warnings, information: :information }

      issues.each_with_object(issues_hash) do |iss, hsh|
        if iss.code == 'code-invalid'
          issues_hash[:information] << "#{issue_location(iss)}: #{iss&.details&.text}"
        elsif DETAILS_FILTER.any? { |filter| filter.match? iss&.details&.text}
          issues_hash[:information] << "#{issue_location(iss)}: #{iss&.details&.text}"
        elsif issue_map.keys.include? iss.severity.to_sym
        issues_hash[issue_map[iss.severity.to_sym]] << "#{issue_location(iss)}: #{iss&.details&.text}"
        end
      end
```

Reply:
I discussed with @arscan and might see if we can do this in a functional way similar to how it was done before...

Might be worth saving a more major refactor for a future PR to get this in quick.